kcron-core/src/commonMain/kotlin/com/ucasoft/kcron/core/builders/Builder.kt

```diff
@@ -65,35 +65,36 @@ class Builder(firstDayOfWeek: WeekDays = WeekDays.Monday) {
         return this
     }
 
-    fun months(type: MonthGroups, value: String) : Builder {
+    fun months(type: MonthGroups, value: String) : Builder<T, D, P> {
         partBuilders.getValue(CronPart.Months).commonBuild(type, value)
         return this
     }
 
-    fun daysOfWeek(type: DayOfWeekGroups, value: String) : Builder {
+    fun daysOfWeek(type: DayOfWeekGroups, value: String) : Builder<T, D, P> {
         if (type != DayOfWeekGroups.Any) {
             days(DayGroups.Any, "?")
         }
         partBuilders.getValue(CronPart.DaysOfWeek).commonBuild(type, value)
         return this
     }
 
-    fun years(type: YearGroups, value: String) : Builder {
+    fun years(type: YearGroups, value: String) : Builder<T, D, P> {
         partBuilders.getValue(CronPart.Years).commonBuild(type, value)
         return this
     }
 
     @DelicateIterableApi
-    fun asIterable(start: LocalDateTime): Iterable<LocalDateTime> {
+    fun asIterable(start: T): Iterable<T> {
+        val internalStart = dateTimeProvider.from(start)
         return Iterable {
             iterator {
-                for (year in (partBuilders.getValue(CronPart.Years) as YearsBuilder).years.filter { y -> y >= start.year }) {
-                    for (month in (partBuilders.getValue(CronPart.Months) as MonthsBuilder).months.filter { m -> m >= start.monthNumber || year > start.year}) {
-                        for (day in calculateDays(year, month, (partBuilders.getValue(CronPart.DaysOfWeek) as DaysOfWeekBuilder).daysOfWeek, (partBuilders.getValue(CronPart.Days) as DaysBuilder).days, start)) {
-                            for (hour in (partBuilders.getValue(CronPart.Hours) as HoursBuilder).hours.filter { h -> h >= start.hour || day > start.dayOfMonth || month > start.monthNumber || year > start.year }) {
-                                for (minute in (partBuilders.getValue(CronPart.Minutes) as MinutesBuilder).minutes.filter { m -> m >= start.minute || hour > start.hour || day > start.dayOfMonth || month > start.monthNumber || year > start.year }) {
-                                    for (seconds in (partBuilders.getValue(CronPart.Seconds) as SecondsBuilder).seconds.filter { s -> s > start.second || minute > start.minute || hour > start.hour || day > start.dayOfMonth || month > start.monthNumber || year > start.year }) {
-                                        yield(LocalDateTime(year, month, day, hour, minute, seconds))
+                for (year in (partBuilders.getValue(CronPart.Years) as YearsBuilder<*>).years.filter { y -> y >= internalStart.year }) {
```

Comment:
The nested for with inside many conditions to filter seems a little bit complex to read.